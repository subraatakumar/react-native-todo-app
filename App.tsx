/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * Generated with the TypeScript template
 * https://github.com/react-native-community/react-native-template-typescript
 *
 * @format
 */

import React from 'react';
import {
  SafeAreaView,
  ScrollView,
  StatusBar,
  StyleSheet,
  Text,
  useColorScheme,
  View,
  Pressable,
  TextInput,
} from 'react-native';
import {ArrowLeft2, ArrowRight2, Verify, Trash} from 'iconsax-react-native';
import {format, addDays} from 'date-fns';
import {TodoObject} from './types';
import AsyncStorage from '@react-native-async-storage/async-storage';

const App = () => {
  type AppData = {
    todos: TodoObject[];
  };

  const [navBarDate, setNavBarDate] = React.useState(new Date());
  const [todos, setTodos] = React.useState<TodoObject[]>([]);
  console.log(todos);
  const [textInputValue, setTextInputValue] = React.useState<string>('');

  const handleRightPress = () => {
    setNavBarDate(addDays(navBarDate, 1));
  };

  const handleLeftPress = () => {
    setNavBarDate(addDays(navBarDate, -1));
  };

  const handleDeletePress = (id: number) => {
    setTodos(todos.filter(item => item.id !== id));
  };

  const storageKey = 'my-app-data';

  const getAppData = async (): Promise<AppData | null> => {
    try {
      const data = await AsyncStorage.getItem(storageKey);

      if (data) {
        return JSON.parse(data);
      }
      return null;
    } catch {
      return null;
    }
  };

  const setAppData = async (newData: AppData) => {
    try {
      await AsyncStorage.setItem(storageKey, JSON.stringify(newData));
    } catch {}
  };

  React.useEffect(() => {
    const getDataFromStorage = async () => {
      const data = await getAppData();

      if (data) {
        setTodos(data.todos);
      }
    };
    getDataFromStorage();
  }, []);

  const handleSubmitPress = () => {
    if (textInputValue === '') return;

    setTodos([...todos, {id: Date.now(), text: textInputValue}]);
    setAppData({todos: todos});
    setTextInputValue('');
  };

  return (
    <SafeAreaView>
      <Pressable onPress={handleLeftPress}>
        <ArrowLeft2 color="#000000" />
      </Pressable>
      <Text>{format(navBarDate, 'dd/MM/yyyy')}</Text>
      <Pressable onPress={handleRightPress}>
        <ArrowRight2 color="#000000" />
      </Pressable>
      <TextInput
        style={styles.textinput}
        value={textInputValue}
        onChangeText={setTextInputValue}
      />
      <Pressable onPress={handleSubmitPress}>
        <ArrowRight2 color="#000000" />
      </Pressable>
      <View>
        {todos.map(singleTodo => (
          <View key={singleTodo.id}>
            <Text>{singleTodo.text}</Text>
            <Pressable onPress={() => handleDeletePress(singleTodo.id)}>
              <Trash color="#000000" />
            </Pressable>
          </View>
        ))}
      </View>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  textinput: {
    borderWidth: 1,
  },
});

export default App;
