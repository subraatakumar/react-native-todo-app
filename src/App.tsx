/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * Generated with the TypeScript template
 * https://github.com/react-native-community/react-native-template-typescript
 *
 * @format
 */

import React, {useEffect} from 'react';
import {SafeAreaView, StyleSheet} from 'react-native';
import {NavBarComponent} from './Components/NavBarComponent';
// import {AppContextProvider} from './AppContextProvider';
import TodoList from './Components/TodoList';
import TodoInput from './Components/TodoInput';
import TodoStore from './store/TodoStore';
import {AppData} from '../types';
import AsyncStorage from '@react-native-async-storage/async-storage';

export const todoStore = new TodoStore();

const App: React.FC = () => {
  useEffect(() => {
    const getAppData = async (): Promise<AppData | null> => {
      //console.log('App data getting started.');
      try {
        const data = await AsyncStorage.getItem(todoStore.storageKey);

        if (data) {
          console.log('Data in local storage', JSON.parse(data));
          todoStore.setTodos(JSON.parse(data).todosData);
        }
        return null;
      } catch {
        return null;
      }
    };
    getAppData();
  }, []);

  return (
    <SafeAreaView style={styles.bodyStyle}>
      {/* <AppContextProvider> */}
      <NavBarComponent />
      <TodoInput />
      <TodoList />
      {/* </AppContextProvider> */}
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  bodyStyle: {
    justifyContent: 'flex-end',
  },
});

export default App;
