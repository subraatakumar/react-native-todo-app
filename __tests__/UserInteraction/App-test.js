import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import App, {todoStore} from '../../src/App';

// Testing User Interaction
test('given empty todoList, user can add an item to it', () => {
  const {
    getByPlaceholderText,
    getByTestId,
    getAllByText,
    getAllByTestId,
    toJSON,
  } = render(<App />);

  fireEvent.changeText(getByPlaceholderText('Input Todo...'), 'testing todo');
  fireEvent.press(getByTestId('addTodoBtn'));

  const testElements = getAllByText('testing todo');
  expect(testElements).toHaveLength(1); // expect 'testing todo' to be on the list
  expect(toJSON()).toMatchSnapshot(); // Another method of snapshot testing
});
