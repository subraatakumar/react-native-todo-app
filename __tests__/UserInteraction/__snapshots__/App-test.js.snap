// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given empty todoList, user can add an item to it 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "justifyContent": "flex-end",
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": 14,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <RNSVGSvgView
        align="xMidYMid"
        bbHeight={32}
        bbWidth={32}
        fill="none"
        focusable={false}
        height={32}
        meetOrSlice={0}
        minX={0}
        minY={0}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            Object {
              "flex": 0,
              "height": 32,
              "width": 32,
            },
          ]
        }
        vbHeight={24}
        vbWidth={24}
        width={32}
        xmlns="http://www.w3.org/2000/svg"
      >
        <RNSVGGroup
          fill={null}
          propList={
            Array [
              "fill",
            ]
          }
        >
          <RNSVGPath
            d="M15 19.92L8.48 13.4c-.77-.77-.77-2.03 0-2.8L15 4.08"
            propList={
              Array [
                "stroke",
                "strokeWidth",
                "strokeLinecap",
                "strokeLinejoin",
                "strokeMiterlimit",
              ]
            }
            stroke={4278190080}
            strokeLinecap={1}
            strokeLinejoin={1}
            strokeMiterlimit={10}
            strokeWidth="1.5"
          />
        </RNSVGGroup>
      </RNSVGSvgView>
    </View>
    <Text
      style={
        Object {
          "color": "#000000",
          "fontFamily": "Roboto",
          "fontSize": 24,
          "fontWeight": "bold",
        }
      }
    >
      12/03/2022
    </Text>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <RNSVGSvgView
        align="xMidYMid"
        bbHeight={32}
        bbWidth={32}
        fill="none"
        focusable={false}
        height={32}
        meetOrSlice={0}
        minX={0}
        minY={0}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            Object {
              "flex": 0,
              "height": 32,
              "width": 32,
            },
          ]
        }
        vbHeight={24}
        vbWidth={24}
        width={32}
        xmlns="http://www.w3.org/2000/svg"
      >
        <RNSVGGroup
          fill={null}
          propList={
            Array [
              "fill",
            ]
          }
        >
          <RNSVGPath
            d="M8.91 19.92l6.52-6.52c.77-.77.77-2.03 0-2.8L8.91 4.08"
            propList={
              Array [
                "stroke",
                "strokeWidth",
                "strokeLinecap",
                "strokeLinejoin",
                "strokeMiterlimit",
              ]
            }
            stroke={4278190080}
            strokeLinecap={1}
            strokeLinejoin={1}
            strokeMiterlimit={10}
            strokeWidth="1.5"
          />
        </RNSVGGroup>
      </RNSVGSvgView>
    </View>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "padding": 14,
      }
    }
  >
    <TextInput
      blurOnSubmit={false}
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Input Todo..."
      style={
        Object {
          "borderRadius": 5,
          "borderWidth": 1,
          "flex": 1,
          "fontFamily": "Roboto",
          "fontSize": 24,
          "paddingLeft": 15,
          "paddingRight": 15,
        }
      }
      value=""
    />
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#BBECA4",
          "borderRadius": 5,
          "justifyContent": "center",
          "marginLeft": 15,
          "padding": 10,
          "width": 53,
        }
      }
      testID="addTodoBtn"
    >
      <RNSVGSvgView
        align="xMidYMid"
        bbHeight={32}
        bbWidth={32}
        fill="none"
        focusable={false}
        height={32}
        meetOrSlice={0}
        minX={0}
        minY={0}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            Object {
              "padding": 10,
            },
            Object {
              "flex": 0,
              "height": 32,
              "width": 32,
            },
          ]
        }
        vbHeight={24}
        vbWidth={24}
        width={32}
        xmlns="http://www.w3.org/2000/svg"
      >
        <RNSVGGroup
          fill={null}
          propList={
            Array [
              "fill",
            ]
          }
        >
          <RNSVGPath
            d="M8.91 19.92l6.52-6.52c.77-.77.77-2.03 0-2.8L8.91 4.08"
            propList={
              Array [
                "stroke",
                "strokeWidth",
                "strokeLinecap",
                "strokeLinejoin",
                "strokeMiterlimit",
              ]
            }
            stroke={4278190080}
            strokeLinecap={1}
            strokeLinejoin={1}
            strokeMiterlimit={10}
            strokeWidth="1.5"
          />
        </RNSVGGroup>
      </RNSVGSvgView>
    </View>
  </View>
  <RCTScrollView>
    <View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#FBFBFB",
              "borderColor": "#FBFBFB",
              "borderRadius": 5,
              "borderWidth": 1,
              "margin": 10,
              "padding": 15,
            },
            Object {
              "elevation": 10,
              "shadowColor": "#000000",
              "shadowOffset": Object {
                "height": 4,
                "width": 4,
              },
              "shadowRadius": 1,
            },
          ]
        }
      >
        <Text
          style={
            Object {
              "color": "#000000",
              "fontFamily": "Roboto",
              "fontSize": 24,
              "paddingBottom": 20,
            }
          }
        >
          testing todo
        </Text>
        <View
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            testID="delete-btn"
          >
            <RNSVGSvgView
              align="xMidYMid"
              bbHeight="45"
              bbWidth="45"
              fill="none"
              focusable={false}
              height="45"
              meetOrSlice={0}
              minX={0}
              minY={0}
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "borderWidth": 0,
                  },
                  Object {
                    "flex": 0,
                    "height": 45,
                    "width": 45,
                  },
                ]
              }
              vbHeight={24}
              vbWidth={24}
              width="45"
              xmlns="http://www.w3.org/2000/svg"
            >
              <RNSVGGroup
                fill={null}
                propList={
                  Array [
                    "fill",
                  ]
                }
              >
                <RNSVGPath
                  d="M21 5.98c-3.33-.33-6.68-.5-10.02-.5-1.98 0-3.96.1-5.94.3L3 5.98M8.5 4.97l.22-1.31C8.88 2.71 9 2 10.69 2h2.62c1.69 0 1.82.75 1.97 1.67l.22 1.3M18.85 9.14l-.65 10.07C18.09 20.78 18 22 15.21 22H8.79C6 22 5.91 20.78 5.8 19.21L5.15 9.14M10.33 16.5h3.33M9.5 12.5h5"
                  propList={
                    Array [
                      "stroke",
                      "strokeWidth",
                      "strokeLinecap",
                      "strokeLinejoin",
                    ]
                  }
                  stroke={4294901760}
                  strokeLinecap={1}
                  strokeLinejoin={1}
                  strokeWidth="1.5"
                />
              </RNSVGGroup>
            </RNSVGSvgView>
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <RNSVGSvgView
              align="xMidYMid"
              bbHeight="45"
              bbWidth="45"
              fill="none"
              focusable={false}
              height="45"
              meetOrSlice={0}
              minX={0}
              minY={0}
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "borderWidth": 0,
                  },
                  Object {
                    "flex": 0,
                    "height": 45,
                    "width": 45,
                  },
                ]
              }
              vbHeight={24}
              vbWidth={24}
              width="45"
              xmlns="http://www.w3.org/2000/svg"
            >
              <RNSVGGroup
                fill={null}
                propList={
                  Array [
                    "fill",
                  ]
                }
              >
                <RNSVGPath
                  d="m8.38 12 2.41 2.42 4.83-4.84"
                  propList={
                    Array [
                      "stroke",
                      "strokeWidth",
                      "strokeLinecap",
                      "strokeLinejoin",
                    ]
                  }
                  stroke={4280394496}
                  strokeLinecap={1}
                  strokeLinejoin={1}
                  strokeWidth="1.5"
                />
                <RNSVGPath
                  d="M10.75 2.45c.69-.59 1.82-.59 2.52 0l1.58 1.36c.3.26.86.47 1.26.47h1.7c1.06 0 1.93.87 1.93 1.93v1.7c0 .39.21.96.47 1.26l1.36 1.58c.59.69.59 1.82 0 2.52l-1.36 1.58c-.26.3-.47.86-.47 1.26v1.7c0 1.06-.87 1.93-1.93 1.93h-1.7c-.39 0-.96.21-1.26.47l-1.58 1.36c-.69.59-1.82.59-2.52 0l-1.58-1.36c-.3-.26-.86-.47-1.26-.47H6.18c-1.06 0-1.93-.87-1.93-1.93V16.1c0-.39-.21-.95-.46-1.25l-1.35-1.59c-.58-.69-.58-1.81 0-2.5l1.35-1.59c.25-.3.46-.86.46-1.25V6.2c0-1.06.87-1.93 1.93-1.93h1.73c.39 0 .96-.21 1.26-.47l1.58-1.35Z"
                  propList={
                    Array [
                      "stroke",
                      "strokeWidth",
                      "strokeLinecap",
                      "strokeLinejoin",
                    ]
                  }
                  stroke={4280394496}
                  strokeLinecap={1}
                  strokeLinejoin={1}
                  strokeWidth="1.5"
                />
              </RNSVGGroup>
            </RNSVGSvgView>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
